cairo/src/lib.rs

```diff
@@ -60,36 +60,77 @@ macro_rules! gvalue_impl {
             }
         }
 
-        impl<'a> glib::value::FromValueOptional<'a> for $name {
-            unsafe fn from_value_optional(v: &'a glib::value::Value) -> Option<Self> {
-                let ptr = glib::gobject_ffi::g_value_get_boxed(v.to_glib_none().0);
-                assert!(!ptr.is_null());
-                from_glib_none(ptr as *mut $ffi_name)
-            }
+        impl glib::value::ValueType for $name {
+            type Type = Self;
         }
 
-        impl glib::value::SetValue for $name {
-            unsafe fn set_value(v: &mut glib::value::Value, s: &Self) {
-                glib::gobject_ffi::g_value_set_boxed(
-                    v.to_glib_none_mut().0,
-                    s.to_glib_none().0 as glib::ffi::gpointer,
-                );
+        impl<'a> glib::value::FromValue<'a> for $name {
+            type Error = glib::value::WrongValueTypeOrNoneError;
+
+            fn check(value: &'a glib::Value) -> Result<(), Self::Error> {
+                glib::value::WrongValueTypeError::check::<$name>(value)?;
```

Comment:
`check()` could in theory be moved into the error type. `WrongValueTypeError` can generically be checked, `WrongValueTypeOrNoneError` can also be generically checked (check that `value.data[0]` is all zeroes). That way it would only have to be implemented once and not for each value type.

The only potential problem here is if we ever want to support types that don't use all-zeroes for the `None` value. @fengalin do you see a problem here with your plans around `gst::ClockTime`?

Could such a type use a new error type specifically for that type? Probably not as that would then conflict with the `impl<T: FromValue<Error = WrongValueTypeOrNoneError>> FromValue for Option<T>`?

Reply:
Easy: add a new trait that does only this check, has a default impl for normal `NULL` types and make `WrongValueTypeOrNoneError<T: ThatNewTrait>` so it can call into it.

Reply:
Leaving the new trait for that for a later time when actually needed.

Reply:
I think I preferred the `check` method on the `FromValue` `trait`. It feels more natural to me to read: `MyType::check(value)` than `WrongValueTypeError::check(value)`. As an ambiguity resolver, that would lead to `<MyType as glib::FromValue>::check(value)`.

We could also implement `check_type::<T: FromValue>() -> Result<(), <T as FromValue>::Error>` on `Value`, so that we could use `value.check_type<MyType>()`.

Reply:
>  think I preferred the check method on the FromValue trait. It feels more natural to me to read: MyType::check(value) than WrongValueTypeError::check(value). As an ambiguity resolver, that would lead to <MyType as glib::FromValue>::check(value).

I agree but that would require code duplication in all implementors (or at least I saw no way to get rid of that), and it's almost never directly called anyway but you go via the `Value` API (also note how `FromValue` is **not** part of the prelude).

> We could also implement check_type::<T: FromValue>() -> Result<(), <T as FromValue>::Error> on Value, so that we could use value.check_type<MyType>().

That's already there as `Value::is::<MyType>()` :)

Reply:
> I agree but that would require code duplication in all implementors (or at least I saw no way to get rid of that), and it's almost never directly called anyway but you go via the Value API (also note how FromValue is not part of the prelude).

I see. Then I have to say that I don't like `trait GetValueError` :), how about `trait ValueTypeChecker`?

Reply:
> That's already there as Value::is::<MyType>() :)

True, but that's more restrictive than `check_type` which can display helpful info when used with `dbg!` or `println!("{}", ...)`. We can still use `value.check_type().is_ok()` to get the boolean behaviour. Also we could use `?` to early exit, etc.

Reply:
> I see. Then I have to say that I don't like trait GetValueError :), how about trait ValueTypeChecker?

Works for me

> True, but that's more restrictive than check_type which can display helpful info when used with dbg! or println!("{}", ...). We can still use value.check_type().is_ok() to get the boolean behaviour. Also we could use ? to early exit, etc.

Do you have a use-case for this? I'd prefer to not add more API surface unless it's actually useful for something.

Reply:
> Do you have a use-case for this?

Not exactly, I was mostly thinking about dropping an { actual, requested } message while debugging. Like when an application developer uses an API, gets a type mismatch and dumps the error to get an insight of the type received.

Reply:
Mostly helping during development, not for production.

Reply:
For those cases, a simple `value.get::<ExpectedType>().unwrap()` should do already :) Let's worry about that then when it's a bit clearer what kind of API would be useful

Reply:
Still, there is my early exit point. With a boolean, you are forced to perform a C style `if value.is() then return ...` when you could just use Rust style `value.check_type()?`  ;)

Reply:
Can you give a somewhat realistic code example where you would do that but would not use `Value::get`?

Reply:
I guess that would be the same use cases as for `is`. But I don't have an example for any off my mind.

Reply:
I don't have one for `is` either, but that existed before :)

Reply:
:) the more I use Rust, the more dislike functions returning a `boolean`. I find them too restrictive.

Reply:
I only ever needed these things in combination with `get()` so...